# Week7 - Decision Tree Lab

* Train-test split
* Train a decison tree model
* Train a random forest model
* Evaluate the models
* Explain findings

In [32]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

df = pd.read_csv('https://raw.githubusercontent.com/msaricaumbc/DS_data/master/ds602/log_reg/employee-turnover-balanced.csv')


In [33]:
from sklearn.preprocessing import LabelEncoder

nominal_cols = ['department', 'gender', 'position', 'married_or_single','frequency_of_travel']
ordinal_cols = ['education', 'satisfaction_with_environment', 'satisfaction_with_job',
                'last_raise_pct', 'last_performance_rating', 'seniority_level']

# one-hot encoding for nominal columns
df = pd.get_dummies(df, columns=nominal_cols)

# label encoding for ordinal columns
label_encoder = LabelEncoder()
for col in ordinal_cols:
    df[col] = label_encoder.fit_transform(df[col])

In [34]:
df.head()

,left_company,age,commuting_distance,education,satisfaction_with_environment,seniority_level,satisfaction_with_job,last_raise_pct,last_performance_rating,total_years_working,...,position_Research Director,position_Research Scientist,position_Sales Executive,position_Sales Representative,married_or_single_Divorced,married_or_single_Married,married_or_single_Single,frequency_of_travel_Non-Travel,frequency_of_travel_Travel_Frequently,frequency_of_travel_Travel_Rarely
0,No,37,16,3,3,1,2,8,0,9,...,0,0,1,0,1,0,0,0,0,1
1,No,39,3,1,2,1,2,4,0,11,...,0,0,0,0,1,0,0,0,0,1
2,No,52,25,3,2,3,3,11,1,31,...,0,0,0,0,0,1,0,0,1,0
3,No,50,1,2,3,1,2,1,0,19,...,0,0,1,0,0,1,0,1,0,0
4,No,44,4,2,3,1,1,1,0,10,...,0,0,0,0,0,0,1,0,0,1


In [35]:
features=[x for x in df.columns if x!='left_company']
X=df[features]
y=df['left_company']

In [36]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


# Create a Decision Tree classifier
model_dt= DecisionTreeClassifier()

# Fit the model to the training data
model_dt.fit(X_train, y_train)

# Make predictions on the test data
y_pred = model_dt.predict(X_test)

# Calculate the accuracy of the model
accuracy = accuracy_score(y_test, y_pred)
accuracy

0.815

## RandomForest Classifier Model

In [37]:
from sklearn.ensemble import RandomForestClassifier


model_rf = RandomForestClassifier(n_estimators=100, random_state=42)
model_rf.fit(X_train, y_train)
y_pred = model_rf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(accuracy)

0.875


we can see both model accuracies, 

accuracy for Decision Tree is 0.81 
accuracy for RandomForest classifier model is 0.87

So, For this particular type of data set, the ensemble approach of random forest classifier model is more accurate than the decision tree model.